protocol/blockrelay/blockrelay.go

```diff
@@ -0,0 +1,274 @@
+	"github.com/pkg/errors"
+	"sync"
+	"time"
+)
+
+var requestedBlocks = &sharedRequestedBlocks{
+	blocks: make(map[daghash.Hash]struct{}),
+}
+
+type sharedRequestedBlocks struct {
+	blocks map[daghash.Hash]struct{}
+	sync.Mutex
+}
+
+func (s *sharedRequestedBlocks) delete(hash *daghash.Hash) {
+	s.Lock()
+	defer s.Unlock()
+	delete(s.blocks, *hash)
+}
+
+func (s *sharedRequestedBlocks) addIfExists(hash *daghash.Hash) (exists bool) {
+	s.Lock()
+	defer s.Unlock()
+	_, ok := s.blocks[*hash]
+	if ok {
+		return true
+	}
+	s.blocks[*hash] = struct{}{}
+	return false
+}
+
+// StartBlockRelay listens to wire.MsgInvRelayBlock messages, requests their corresponding blocks if they
+// are missing, adds them to the DAG and propagates them to the rest of the network.
+func StartBlockRelay(msgChan <-chan wire.Message, peer *peerpkg.Peer, netAdapter *netadapter.NetAdapter, router *netadapter.Router,
+	dag *blockdag.BlockDAG) error {
+
+	invsQueue := make([]*wire.MsgInvRelayBlock, 0)
+	for {
+		shouldStop, err := handleInv(msgChan, netAdapter, router, peer, dag, invsQueue)
+		if err != nil {
+			return err
+		}
+		if shouldStop {
+			return nil
+		}
+	}
+}
+
+func handleInv(msgChan <-chan wire.Message, netAdapter *netadapter.NetAdapter, router *netadapter.Router,
+	peer *peerpkg.Peer, dag *blockdag.BlockDAG, invsQueue []*wire.MsgInvRelayBlock) (shouldStop bool, err error) {
+
+	inv, shouldStop, err := readInv(msgChan, &invsQueue)
+	if err != nil {
+		return false, err
+	}
+	if shouldStop {
+		return true, nil
+	}
+
+	if dag.IsKnownBlock(inv.Hash) {
+		if dag.IsKnownInvalid(inv.Hash) {
+			return false, errors.Errorf("sent inv of invalid block %s",
+				inv.Hash)
+		}
+		return false, nil
+	}
+
+	requestQueue := []*daghash.Hash{inv.Hash}
+	requestQueueSet := map[daghash.Hash]struct{}{
+		*inv.Hash: {},
+	}
+	pendingBlocks := map[daghash.Hash]struct{}{}
+
+	// In case the function closes earlier than expected, we wanna make sure requestedBlocks is
+	// clean from any pending blocks.
+	defer deleteFromRequestedBlocks(pendingBlocks)
+	for len(requestQueue) > 0 {
+		shouldStop, err := requestBlocks(netAdapter, router, peer, msgChan, dag, &invsQueue,
+			&requestQueue, requestQueueSet)
+		if err != nil {
+			return false, err
+		}
+		if shouldStop {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func readInv(msgChan <-chan wire.Message,
+	invsQueue *[]*wire.MsgInvRelayBlock) (inv *wire.MsgInvRelayBlock, shouldStop bool, err error) {
+
+	if len(*invsQueue) > 0 {
+		inv, *invsQueue = (*invsQueue)[0], (*invsQueue)[1:]
+		return inv, false, nil
+	}
+
+	for {
+		msg, isOpen := <-msgChan
+		if !isOpen {
+			return nil, true, nil
+		}
+
+		inv, ok := msg.(*wire.MsgInvRelayBlock)
+		if ok {
+			return inv, false, nil
+		}
+
+		return nil, false, errors.Errorf("unrequested %s message in the block relay flow", msg.Command())
+	}
+}
+
+// readMsgBlock returns the next msgBlock in msgChan, and populates invsQueue in any inv messages that arrives between.
```

Comment:
Please rephrase: "populates invsQueue with any inv messages that meanwhile arrive"